### PR TITLE
Fix per-frame allocations in `BeatmapCarousel`

### PR DESCRIPTION
**Pull request by peppy**
| Before | After |
| :---: | :---: |
| JetBrains Rider 2024-09-04 at 04 54 23 | JetBrains Rider 2024-09-04 at 05 31 06 |

### CODE REVIEW

File: osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmapSet.cs [hunk 1]
```diff
@@ -51,9 +51,7 @@ public partial class DrawableCarouselBeatmapSet : DrawableCarouselItem, IHasCont
         [Resolved]
         private IBindable<RulesetInfo> ruleset { get; set; } = null!;
 
-        public IEnumerable<DrawableCarouselItem> DrawableBeatmaps => beatmapContainer?.IsLoaded != true ? Enumerable.Empty<DrawableCarouselItem>() : beatmapContainer.AliveChildren;
-
-        private Container<DrawableCarouselItem>? beatmapContainer;
+        public Container<DrawableCarouselItem>? Beatmaps;
```
Comment (smoogipoo) [inline osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmapSet.cs:54]: I think I'd prefer this to be private and exposed as an `IReadOnlyList`.

Comment (peppy) [reply osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmapSet.cs:54]: Just to confirm, this means using the `for` loop as you proposed?

Comment (smoogipoo) [reply osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmapSet.cs:54]: Yeah.

File: osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmapSet.cs [hunk 2]
```diff
@@ -51,7 +51,7 @@ public partial class DrawableCarouselBeatmapSet : DrawableCarouselItem, IHasCont
         [Resolved]
         private IBindable<RulesetInfo> ruleset { get; set; } = null!;
 
-        public IEnumerable<DrawableCarouselItem> DrawableBeatmaps => beatmapContainer?.IsLoaded != true ? Enumerable.Empty<DrawableCarouselItem>() : beatmapContainer.AliveChildren;
+        public IReadOnlyList<DrawableCarouselItem> DrawableBeatmaps => beatmapContainer?.IsLoaded != true ? Array.Empty<DrawableCarouselItem>() : beatmapContainer;
```
Comment (bdach) [inline osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmapSet.cs:54]: Can't immediately tell if the `AliveChildren` to `Children` swap here is correct but I don't think we do any lifetime games in song select so I *think* it should be

**Review by bdach — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]

File: osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmapSet.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmapSet.cs:54]: Yeah it seemed fine. Worst case it's just setting some values when not required, but it's worth the alloc saving.

**Pull request merged**